lms/tasks/email_digests.py

```diff
@@ -1,13 +1,70 @@
 import logging
-from datetime import datetime
+from datetime import datetime, timedelta, timezone
 from typing import List, Optional
 
+from sqlalchemy import select
+
+from lms.models import ApplicationInstance, AssignmentMembership, LTIRole, User
 from lms.services import DigestService
 from lms.tasks.celery import app
 
 LOG = logging.getLogger(__name__)
 
 
+@app.task
+def send_instructor_email_digest_tasks(batch_size):
+    """
+    Generate and send instructor email digests.
+
+    The email digests will cover activity that occurred in the time period from
+    5AM UTC yesterday morning to 5AM UTC this morning.
+
+    Emails will be sent to all instructors who're participating in the feature
+    and who have digest activity in the time period.
+
+    5AM UTC is chosen because it equates to midnight EST. Most of our target
+    users for this feature are in the EST timezone and we want each email
+    digests to cover "yesterday" (midnight to midnight) EST.
+
+    EST is 5 hours behind UTC (ignoring daylight savings for simplicity: we
+    don't need complete accuracy in the timing).
+    """
+    now = datetime.now(timezone.utc)
+    updated_before = datetime(year=now.year, month=now.month, day=now.day, hour=5)
+    updated_after = updated_before - timedelta(days=1)
+    updated_before = updated_before.isoformat()
+    updated_after = updated_after.isoformat()
+
+    with app.request_context() as request:  # pylint:disable=no-member
+        with request.tm:
+            h_userids = request.db.scalars(
+                select(User.h_userid)
+                .select_from(User)
+                .distinct()
+                .join(ApplicationInstance)
+                .join(AssignmentMembership)
+                .join(LTIRole)
+                .where(
+                    ApplicationInstance.settings["hypothesis"][
+                        "instructor_email_digests_enabled"
+                    ].astext
+                    == "true",
+                    LTIRole.type == "instructor",
+                )
+            ).all()
```

Comment:
We talked about moving this DB query to a generic place, maybe `UserService`. I could do that in a follow-up PR but I'm not sure if it's worth it: the DB query is very specific to this task, getting just the distinct `h_userid`'s for user belonging to application instances with the `hypothesis.instructor_email_digests_enabled` setting on and who are instructors. By the time we've create a generic `get_users()` method in some service with enough parameters to recreate this query I think we might just be reinventing the sqlalchemy API.

Reply:
Yeah. I think moving the membership from assignments to courses is the right thing to do but it won't really make the query any easier, you'll just have to join to `GroupingMembership` instead:

```python
.select_from(User)
.distinct()
.join(ApplicationInstance)
.join(GroupingMembership)
.join(LTIRole)
```

I think the need to join to `ApplicationInstance` may be unavoidable since we're using an `ApplicationInstance` feature flag. Moving the feature flag into `Organization` would just mean needing to join to `Organization`, I think you'd actually still need to join to `ApplicationInstance` to get to the `Organization` and then you'd need to deal with hierarchical organizations, so that's actually much worse.

I think having to join to `AssignmentMembership` (or `GroupingMembership` if we fixed it) is also unavoidable: this is just part of the external domain model of the LMS app, users aren't simply instructors rather they're either an instructor or not within the context of each course.

I wonder if we've over-complicated things with the `LTIRole` table? It might be nice to be able to do something like:

```python
.select_from(User)
.distinct()
.join(ApplicationInstance)
.join(GroupingMembership)
.where(GroupingMembership.role == "instructor")
```

It's only a minor simplification though.

Anyway yeah maybe there'll be an opportunity to move this into a service later.